### ISSUE TITLE
Is there a way to disable plugins included in the project's config?

### ISSUE THREAD

**Issue by davidmh** [pmizio/typescript-tools.nvim#78]
Hi @pmizio great work with this plugin, I've tried it in a couple of projects and it works flawlessly.

The only problem I'm facing at the moment happens to be in the main project at my job. It's a big project with around 5k files. When I open a file, the LSP client starts just fine, I can find references and get the hover window just fine.

The problem comes when I make any sort of changes, that is, adding a comment, renaming a function, etc. Then the LSP client becomes unresponsive for up to 10 seconds, then it goes back to normal.

I've narrowed down the problem to a Typescript plugin: `typescript-plugin-css-modules` with less: https://github.com/dancon/less-plugin-aliases#work-with-webpack-and-typscript-plugin-css-modules

If I remove that plugin declaration from `tsconfig.json` and restart the LSP client, the problem goes away.

That plugin is a nice thing to have but I can live without it if it means I can have a great LSP client.

Is there a way to ignore plugins defined in the project's configuration?

---

For context, I don't think the problem is related to `typescript-plugin-css-modules` or `less-plugin-aliases` since my coworkers are using the same configuration in VSCode and they don't see this problem. It also works fine with `typescript-language-server`.

My current driver is [`vtsls`](https://github.com/yioneko/vtsls) which performs way better than `typescript-language-server`, but doesn't support plugins, which could explain why it doesn't have perf issues.

Unfortunately, that project is not publicly available, so I can't share its details.

If it helps, this is the minimal config file I used to try it out:

<details>
<summary>minimal_init.lua</summary>

```lua
local M = {}

function M.root(root)
    local f = debug.getinfo(1, "S").source:sub(2)
    return vim.fn.fnamemodify(f, ":p:h:h") .. "/" .. (root or "")
end

---@param plugin string
function M.load(plugin)
    local name = plugin:match(".*/(.*)")
    local package_root = M.root(".minimal/site/pack/deps/start/")
    if not vim.loop.fs_stat(package_root .. name) then
        print("Installing " .. plugin)
        vim.fn.mkdir(package_root, "p")
        vim.fn.system({
            "git",
            "clone",
            "--depth=1",
            "https://github.com/" .. plugin .. ".git",
            package_root .. "/" .. name,
        })
    end
end

function M.setup()
    vim.cmd([[set runtimepath=$VIMRUNTIME]])
    vim.opt.runtimepath:append(M.root())
    vim.opt.packpath = { M.root(".minimal/site") }
    vim.env.XDG_CONFIG_HOME = M.root(".minimal/config")
    vim.env.XDG_DATA_HOME = M.root(".minimal/data")
    vim.env.XDG_STATE_HOME = M.root(".minimal/state")
    vim.env.XDG_CACHE_HOME = M.root(".minimal/cache")

    M.load("nvim-lua/plenary.nvim")
    M.load("neovim/nvim-lspconfig")
    M.load("j-hui/fidget.nvim")
    M.load("pmizio/typescript-tools.nvim")
    M.load("catppuccin/nvim")

    require("catppuccin").setup({ flavour = "macchiato" })
    vim.cmd.colorscheme("catppuccin")
    require("fidget").setup({})
    require("typescript-tools").setup({
        tsserver_plugins = {},
    })

    vim.api.nvim_create_augroup("lsp-config", { clear = true })
    vim.api.nvim_create_autocmd("LspAttach", {
        group = "lsp-config",
        callback = M.lsp_mappings,
    })
end

M.lsp_mappings = function()
    vim.api.nvim_buf_set_option(0, "omnifunc", "v:lua.vim.lsp.omnifunc")
    local opts = { buffer = true, silent = true }
    vim.keymap.set("n", "K", vim.lsp.buf.hover, opts)
    vim.keymap.set("n", "gd", vim.lsp.buf.definition, opts)
    vim.keymap.set("n", "<leader>lf", vim.lsp.buf.references, opts)
    vim.keymap.set("n", "<leader>li", vim.lsp.buf.implementation, opts)
    vim.keymap.set("n", "<leader>lt", vim.lsp.buf.type_definition, opts)
    vim.keymap.set("n", "<leader>la", vim.lsp.buf.code_action, opts)
    vim.keymap.set("n", "<leader>lr", vim.lsp.buf.rename, opts)
end

vim.o.swapfile = false
_G.__TEST = true

M.setup()
```
</details>

**Comment by pmizio** [comment]
From top of my head probably not but I'll double-check that. But we can try to trace problem. You can enable tsserver logs using [this](https://github.com/pmizio/typescript-tools.nvim/blob/1a1fc8fcc498ade3adf8b251e930abb2eb5b7983/lua/typescript-tools/config.lua#L49) property and they live in os temp folder as tsserver_syntax.log and tsserver_semantic.log files. In logs you can find how long requests take.
I also try to reproduce this I also have in my job large  project with css modules but we generate declarations instead using ts plugin, maybe to test I install it and luckly enough problem also show up.

About VSCode working - it may just don't load plugin at all, because by default it don't use workspace typescript installation.  
Your colleagues may use default behaviour which is load bundled typescript in vscode installation without any plugin.
You may ask them about this config. Here are instructions: https://code.visualstudio.com/docs/typescript/typescript-compiling#_using-the-workspace-version-of-typescript

Basically if after running `Select TypeScript Version` from vscode command palette they don't have `Use Workspace Version` checked they actually don't have this plugin loaded.

**Comment by yioneko** [comment]
@davidmh Sorry to bother you here, but I want to clarify that vtsls does support plugins, the reason that it doesn't activate in your case is that the server uses bundled typescript version by default instead of workspace version, just the same as vscode. You can switch to workspace version by the dedicated command `select_ts_version` as in vscode or through configuration. I saw someone also has this confusion, as unclear docs about this is definitely my fault.

**Comment by pmizio** [comment]
In this case probably default vscode setting is blessing because PROBABLY this typescript cssmodules plugin ruin experience. But I wait with final judgement to time @davidmh provide more information how vscode behaves on workspace tsserver and/or what is situation in logs.

**Comment by davidmh** [comment]
> You can switch to workspace version by the dedicated command select_ts_version as in vscode or through configuration

Oh that's great to know, I'll give it a try today.

> provide more information how vscode behaves on workspace tsserver and/or what is situation in logs.

I confirmed with my coworkers that the css modules plugin does work on vscode and it doesn't show the same problem I see in neovim.

As for the logs, this is the only relevant entry that shows a slow process: 

```
Info 4201 [08:07:58.632] -----------------------------------------------
Info 4201 [08:07:58.632] Open files: 
Info 4201 [08:07:58.632] 	FileName: <REDACTED>.tsx ProjectRootPath: undefined
Info 4201 [08:07:58.632] 		Projects: <REDACTED>/tsconfig.json
Perf 4201 [08:07:58.632] 7::updateOpen: elapsed time (in milliseconds) 12521.3447
Info 4202 [08:07:58.632] response:
    {"seq":0,"type":"response","command":"updateOpen","request_seq":7,"success":true,"performanceData":{"updateGraphDurationMs":10822.790082991123},"body":true}
Info 4203 [08:07:58.635] request:
    {"type":"request","arguments":{"file":"<REDACTED>.tsx","preferences":{"includeInlayVariableTypeHints":false,"includeInlayVariableTypeHintsWhenTypeMatchesName":false,"includeInlayPropertyDeclarationTypeHints":false,"includeInlayFunctionLikeReturnTypeHints":false,"includeInlayEnumMemberValueHints":false,"providePrefixAndSuffixTextForRename":true,"allowRenameOfImportPath":true,"quotePreference":"auto","importModuleSpecifierEnding":"auto","jsxAttributeCompletionStyle":"auto","allowTextChangesInNewFiles":true,"includeAutomaticOptionalChainCompletions":true,"provideRefactorNotApplicableReason":true,"generateReturnInDocTemplate":true,"includeCompletionsForImportStatements":true,"includeCompletionsWithSnippetText":true,"includeCompletionsWithClassMemberSnippets":true,"includeCompletionsWithObjectLiteralMethodSnippets":true,"useLabelDetailsInCompletionEntries":true,"allowIncompleteCompletions":true,"displayPartsForJSDoc":true,"disableLineTextInReferences":true,"includeInlayParameterNameHints":"none","includeInlayParameterNameHintsWhenArgumentMatchesName":false,"includeInlayFunctionParameterTypeHints":false},"formatOptions":{"insertSpaceAfterFunctionKeywordForAnonymousFunctions":true,"insertSpaceBeforeFunctionParenthesis":false,"insertSpaceAfterOpeningAndBeforeClosingNonemptyParenthesis":false,"insertSpaceAfterOpeningAndBeforeClosingNonemptyBrackets":false,"indentSize":2,"convertTabsToSpaces":true,"tabSize":2,"insertSpaceAfterOpeningAndBeforeClosingJsxExpressionBraces":false,"insertSpaceAfterTypeAssertion":false,"placeOpenBraceOnNewLineForFunctions":false,"placeOpenBraceOnNewLineForControlBlocks":false,"semicolons":"ignore","indentSwitchCase":true,"insertSpaceAfterOpeningAndBeforeClosingTemplateStringBraces":false,"insertSpaceAfterOpeningAndBeforeClosingEmptyBraces":true,"insertSpaceAfterOpeningAndBeforeClosingNonemptyBraces":true,"newLineCharacter":"\n","insertSpaceAfterCommaDelimiter":true,"insertSpaceAfterConstructor":false,"insertSpaceAfterSemicolonInForStatements":true,"insertSpaceBeforeAndAfterBinaryOperators":true,"insertSpaceAfterKeywordsInControlFlowStatements":true}},"seq":11,"command":"configure"}
Info 4204 [08:07:58.635] Host configuration update for file <REDACTED>.tsx
Info 4205 [08:07:58.635] response:
    {"seq":0,"type":"response","command":"configure","request_seq":11,"success":true}
Perf 4206 [08:07:58.635] 11::configure: async elapsed time (in milliseconds) 0.3023
Info 4207 [08:07:58.635] request:
    {"type":"request","arguments":{"openFiles":[{"file":"<REDACTED>.tsx","fileContent":"<REDACTED>","scriptKindName":"TSX"}],"closedFiles":[],"changedFiles":[]},"seq":11,"command":"updateOpen"}
Info 4208 [08:07:58.640] Project '<REDACTED>/tsconfig.json' (Configured)
Info 4208 [08:07:58.640] 	Files (5194)
```

Let me know if you're looking of a specific event.

**Comment by pmizio** [comment]
If I interpret logs correctly you presenting me few first events after server start and after you open first file project start to index which can take in large projects quite a few seconds. To the time server stop indexing all requests in the meantime are queued and flushed when indexing stop. After that process LSP should be responsive all the time.
Do you use some LSP progress indicator like [fidget.nvim](https://github.com/j-hui/fidget.nvim)? It show when LSP indexing things then when it disappear LSP should be interactive.
As far as I know vscode in time of indexing create some temporary server which serve minimal features then after indexing requests are routed to main server. But this plugin not implement this sorry.

At least this is what I see from this part of log you show me. Can you @davidmh confirm/deny this?

**Comment by davidmh** [comment]
> If I interpret logs correctly you presenting me few first events after server start and after you open first file project start to index which can take in large projects quite a few seconds. To the time server stop indexing all requests in the meantime are queued and flushed when indexing stop. After that process LSP should be responsive all the time.

No, the part of the log I posted happened after the project was fully indexed. The lag happens *after* I make any changes to the codebase.

> Do you use some LSP progress indicator like [fidget.nvim](https://github.com/j-hui/fidget.nvim)? It show when LSP indexing things then when it disappear LSP should be interactive.
As far as I know vscode in time of indexing create some temporary server which serve minimal features then after indexing requests are routed to main server. But this plugin not implement this sorry.

I do use fidget, but it only shows the initial indexing, nothing after that.

> Like you it is almost identical, and after this first req which took about ~12s rest of requests are 100% interactive for me.
I also noticed this ts plugin throw some errors in tsserver log but everything still after indexing works fine and snappy like without this ts plugin.

Yeah, removing the plugin does make the LSP client super snappy.

FWIW, I just tested the settings recommended by @yioneko using `vtsls` here: https://github.com/davidmh/dot-files/blob/main/nvim/fnl/own/plugin/lsp.fnl#L109-L110

I can tell that the problem is definitely related to that CSS modules plugin, because now I get the typings for the styles, but the client does show a small lag after any updates.

It doesn't seem to be as bad though. I'll try it for a few days and see how it feels with actual work.

Feel free to close this issue for now, since the problem seems to be related to the performance of the typescript plugin, not the LSP client.
